packages/components/docs/immer.md

```diff
@@ -0,0 +1,235 @@
+will either fail to modify or throw a run-time error.
+
+Now, let's actually make an update to the immutable instance with Immer:
+
+```ts
+import { produce } from 'immer';
+
+// Create the initial instance
+const circle = produce(new Circle(5), () => {});
+
+// Mutate and copy from `produce`
+const newCircle = produce(circle, (draft) => {
+    draft.radius = 10;
+});
+console.log(circle.radius);    // 5
+console.log(newCircle.radius); // 10
+```
+
+The instance is now immutable, a mutated copy can be made via `produce`, and we have run-time safety via `Object.freeze`.
+
+#### Compile-time safety
+
+The main advantage of using TypeScript is that it compiles your code to ensure correctness before you run it. Additionally,
+the run-time safety is great but it still may be something that only occurs in production if say our test coverage misses
+a code path.
+
+To get started, let's first declare all the properties on the `Circle` class as [`readonly`](https://www.typescriptlang.org/docs/handbook/classes.html#readonly-modifier).
+
+```ts
+import { immerable } from 'immer';
+
+class Circle {
+    [immerable] = true;
+
+    readonly radius: number;
+
+    constructor(r: number) {
+        this.radius = r;
+    }
+}
+```
+
+The `radius` property is now read-only so if we attempt to modify it directly we receive an error:
+
+```ts
+import { produce } from 'immer';
+
+let circle = new Circle(5);
+circle.radius = 5; // Error: TS2540: Cannot assign to 'radius' because it is a read-only property.
+
+// Same for the produced version
+let circle = produce(new Circle(5), () => {});
+circle.radius = 5; // Error: TS2540: Cannot assign to 'radius' because it is a read-only property.
+```
+
+This gives us compile-time safety against invalid writes. The next feature we can use is the `Draft` utility from
+Immer. To quote the docs:
+
+> For curried reducers, the type is inferred from the first argument of recipe function, so make sure to type it.
+The `Draft` utility type can be used if the state argument type is immutable.
+
+This will help ensure the passed in type is immutable.
+
+```ts
+import { Draft, produce } from 'immer';
+
+// Without "Draft"
+let circle = produce(new Circle(5), (draft: Circle) => {
+    draft.radius = 10; // Error: TS2540: Cannot assign to 'radius' because it is a read-only property.
+});
+
+// With "Draft"
+let circle = produce(new Circle(5), (draft: Draft<Circle>) => {
+    draft.radius = 10; // OK!
+});
+```
+
+If you'd like to have your class instance's be immutable without requiring use of `produce` you can directly call
+`Object.freeze` at the end of the constructor.
+
+```ts
+import { immerable } from 'immer';
+
+class Circle {
+    [immerable] = true;
+
+    readonly radius: number;
+
+    constructor(r: number) {
+        this.radius = r;
+
+        // Optionally, freeze the instance. Still works with produce but doesn't require it to have
+        // an immutable instance via construction.
+        Object.freeze(this);
```

Comment:
So just out of curiosity and possibly for clarification in the docs, do you even need readonly properties if you're going to do Object.freeze in constructor?  If not, could you then just do Object.freeze in the constructor and not have to use Draft<> wrapper in produce?

Reply:
Oh I see Object.freeze is a runtime thing.  Which I see is necessary because it is surprisingly easy to get around readonly with a slightly indirect reference to it.  Maybe we should call out that this example is the template we should be using.

// readonly fail
let circle = new Circle(5);
let r = "radius";
circle[r] = 7;

console.log(circle.radius);  // 7

Reply:
#### readonly
Using `readonly` is intended to help prevent illicit behavior but, as you noted, the TS compiler can't catch everything. Minor thing, but if you changed it to `const r = "radius"` then TS does pick up the type and will throw an error.

#### Object.freeze
Unfortunately, `Object.freeze` also doesn't suffice in more complex cases because it is not a "deep freeze" Example:

```js
'use strict';
let x = Object.freeze({ y: { z: 10 } });
x.y = {};            // TypeError: Cannot assign to read only property 'y' of object '#<Object>'
x.y.z = 20;          // OK! Because 'y' is not frozen.
console.log(x.y.z);  // 20
```

There isn't really a clean way (that I can think of) to have a deeply immutable class instance using Immer via the class' constructor only. It's kind of the inverse of the problem we have with ImmutableJS where ImmutableJS the instance is already immutable after the required call to `super()`. Ideally, I'd like to have `new Circle(5)` return me a deep-freezed immutable instance.